http/src/handler.rs

```diff
@@ -70,87 +63,227 @@ impl<M: Metadata, S: Middleware<M>> server::Handler<HttpStream> for ServerHandle
+			Ready(handler) => (handler, true),
+			NotReady(handler) => (handler, false),
 		}
 	}
 }
+enum RpcHandlerState<M> {
+	ReadingHeaders {
+		request: server::Request,
+		cors_domains: CorsDomains,
+		continue_on_invalid_cors: bool,
+	},
+	ReadingBody {
+		body: hyper::Body,
+		request: Vec<u8>,
+		metadata: M,
+	},
+	Writing(Response),
+	Waiting(BoxFuture<Option<String>, ()>),
+	Done,
+}
+
+impl<M> fmt::Debug for RpcHandlerState<M> {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+		use self::RpcHandlerState::*;
 
-impl<M: Metadata, S: Middleware<M>> DerefMut for Handler<M, S> {
-	fn deref_mut(&mut self) -> &mut Self::Target {
 		match *self {
-			Handler::Rpc(ref mut handler) => handler,
-			Handler::Error(ref mut handler) => handler,
-			Handler::Middleware(ref mut response) => &mut **response,
+			ReadingHeaders {..} => write!(fmt, "ReadingHeaders"),
+			ReadingBody {..} => write!(fmt, "ReadingBody"),
+			Writing(ref res) => write!(fmt, "Writing({:?})", res),
+			Waiting(_) => write!(fmt, "Waiting"),
+			Done => write!(fmt, "Done"),
 		}
 	}
 }
 
 pub struct RpcHandler<M: Metadata, S: Middleware<M>> {
-	cors_domains: Option<Vec<cors::AccessControlAllowOrigin>>,
 	jsonrpc_handler: Rpc<M, S>,
-	metadata: Option<M>,
-	control: Control,
-	request: Request,
-	response: Response,
-	continue_on_invalid_cors: bool,
-	/// Asynchronous response waiting to be moved into `response` field.
-	waiting_sender: mpsc::Sender<Response>,
-	waiting_response: mpsc::Receiver<Response>,
+	state: RpcHandlerState<M>,
+	is_options: bool,
+	cors_header: cors::CorsHeader<header::AccessControlAllowOrigin>,
+}
+
+impl<M: Metadata, S: Middleware<M>> Future for RpcHandler<M, S> {
+	type Item = server::Response;
+	type Error = hyper::Error;
+
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		let new_state = match mem::replace(&mut self.state, RpcHandlerState::Done) {
+			RpcHandlerState::ReadingHeaders { request, cors_domains, continue_on_invalid_cors, } => {
+				// Read cors header
+				self.cors_header = utils::cors_header(&request, &cors_domains);
+				self.is_options = *request.method() == Method::Options;
+				// Read other headers
+				RpcPollState::Ready(self.read_headers(request, continue_on_invalid_cors))
+			},
+			RpcHandlerState::ReadingBody { body, request, metadata, } => {
+				self.read_body(body, request, metadata)?
+			},
+			RpcHandlerState::Waiting(mut waiting) => {
+				match waiting.poll() {
+					Ok(Async::Ready(response)) => {
+						RpcPollState::Ready(RpcHandlerState::Writing(match response {
+							// Notification, just return empty response.
+							None => Response::ok(String::new()),
+							// Add new line to have nice output when using CLI clients (curl)
+							Some(result) => Response::ok(format!("{}\n", result)),
+						}.into()))
+					},
+					Ok(Async::NotReady) => RpcPollState::NotReady(RpcHandlerState::Waiting(waiting)),
+					Err(_) => RpcPollState::Ready(RpcHandlerState::Writing(Response::internal_error())),
```

Comment:
It's not really `hyper::Error` that we are getting here waiting is `BoxFuture<Option<String>, ()>` and it's an error coming from `jsonrpc_core` which (in theory) can never fail. That's why we return `internal_error` here.